clar2wasm/src/tools.rs

```diff
@@ -392,6 +392,36 @@ pub fn crosscheck_compare_only(snippet: &str) {
     let compiled = evaluate(snippet);
     let interpreted = interpret(snippet);
 
+    // to avoid false positives when both the compiled and interpreted fail,
+    // we don't allow failures in these tests
+
+    if let Err(e) = compiled {
+        panic!("Compiled snippet failed: {:?}", e);
+    }
+
+    assert_eq!(
+        compiled, interpreted,
+        "Compiled and interpreted results diverge! {}\ncompiled: {:?}\ninterpreted: {:?}",
+        snippet, &compiled, &interpreted
+    );
+}
+
+pub fn crosscheck_compare_only_with_expected_error<E: Fn(&Error) -> bool>(
+    snippet: &str,
+    expected: E,
+) {
+    let compiled = evaluate(snippet);
+    let interpreted = interpret(snippet);
+
+    // to avoid false positives when both the compiled and interpreted fail,
+    // we don't allow failures in these tests
+
+    if let Err(e) = &compiled {
+        if !expected(e) {
+            panic!("Compiled snippet failed with unexpected error: {:?}", e);
+        }
+    }
```

Comment:
If we compare the expected error here before the assertion, we might have an issue where the compiled and interpreted are equal, but we wrote a wrong assertion for the failure case.

Also, Nothing happens in the `Ok` case. Is that the point of this function, to allow either `Ok` or `Err` values? 
Completely hypothetical case: I use this function to check that my snippet provokes an error, but both my snippets actually returns an `Ok`. Then this function will succeed, even if I wanted it to fail.

Reply:
This is for example where we property-test arbitrary additions/multiplications where we can narrow down the allowed errors to arithmetic ones, catching other issues/regressions.

Reply:
as is, if you want a specific failure you just use regular crosscheck.